src/prover_disk.hpp

```diff
@@ -241,6 +242,32 @@ class DiskProver {
     std::vector<uint64_t> table_begin_pointers;
     std::vector<uint64_t> C2;
 
+    // Using this method instead of simply seeking will prevent segfaults that would arise when
+    // continuing the process of looking up qualities.
+    void safeSeek(std::ifstream& disk_file, uint64_t seek_location) {
+        if ((disk_file.rdstate() & std::ifstream::goodbit) != 0) {
+            throw std::invalid_argument("safeSeed: provided input stream goodbit != 0.");
+        }
+
+        disk_file.seekg(seek_location);
+
+        if ((disk_file.rdstate() & std::ifstream::goodbit) != 0) {
+            std::cout << "goodbit, failbit, badbit, eofbit: " <<   (disk_file.rdstate() & std::ifstream::goodbit) <<  (disk_file.rdstate() & std::ifstream::failbit) << (disk_file.rdstate() & std::ifstream::badbit) << (disk_file.rdstate() & std::ifstream::eofbit) << std::endl;
+            throw std::runtime_error("goodbit!=0 after seeking to" + std::to_string(seek_location));
+        }
+    }
+
+    void safeRead(std::ifstream& disk_file, uint8_t* target, uint64_t size) {
+        if ((disk_file.rdstate() & std::ifstream::goodbit) != 0) {
```

Comment:
Probably it's extraneous to check the state before `read`. Also see the next comment about `goodbit` being 0.